simpleui-server/src/SuiZMQ.ts

```diff
@@ -0,0 +1,150 @@
+ * Created by Zack Beucler
+ * 
+ * potenitally add a send timeout ZMQ_SNDTIMEO
+ *                      http://api.zeromq.org/3-0:zmq-setsockopt
+ *      might be good to add because zmq will wait forever if not told otherwise
+ * 
+ * 
+ ****************************** NOTE ******************************
+ *
+ *  An http request can only request data from a single zmq socket
+ *
+ ******************************************************************
+ */
+
+
+var _zmq = require('zeromq');
+import {ServerUtil} from './server-util';
+import {Logger, LogLevel} from './server-logger';
+import { SuiData } from './sui_data';
+import { Queue } from './queue';
+
+
+export class ZMQ_Socker_Wrapper {
+    hostname: string;
+    timeout: number;
+    port: number;
+    socket: any;
+    http_queue: Queue;
+    
+
+    constructor(hostname: string='svcmachineapps', port: number, timeout: number=500) {
+        this.hostname = hostname;
+        this.timeout = timeout;
+        this.port = port;
+        this.http_queue = new Queue();
+
+        try {
+            this.socket = _zmq.socket('req');
+            this.socket.connect_timeout = timeout;
+            this.connect(this.port);
+
+            this.socket.on('message', (msg) => {
+                const raw_zmq_data = msg.toString();
+                const zmq_data = SuiData.addXmlStatus(raw_zmq_data);
+                Logger.log(LogLevel.DEBUG, `Recieved ZMQ message at port ${this.port}: ${zmq_data.substring(0, 105)}`);
+                // get response + request from queue
+                let [res, req] = this.http_queue.dequeue();
+                // send response
+                SuiData.sendResponse(req, res, zmq_data);
+            });
+
+            this.socket.on('error', (zmqErr) => {
+                const zmq_data = ServerUtil.getServerError('ZMQ_ERROR', '{{ERROR}}', zmqErr);
+                Logger.log(LogLevel.ERROR, `ZMQ socket at port ${this.port} got error: ${zmqErr}`);
+                // get response + request from queue
+                let [res, req] = this.http_queue.dequeue();
+                // send response
+                SuiData.sendResponse(req, res, zmq_data);
+            });
+
+        } catch (err) {
+            Logger.log(LogLevel.ERROR, `Could not create ZMQ socket at port ${this.port} got error: ${err}`)
+        }
+    }
+
+    connect(port: number) {
+        try {
+            Logger.log(LogLevel.VERBOSE, `ZMQ connecting to --->  tcp://${this.hostname}:${port}`)
+            this.socket.connect(`tcp://${this.hostname}:${port}`)
+
+        } catch (e) {
+            Logger.log(LogLevel.ERROR, `Could not connect to tcp://${this.hostname}:${port}`);
+        }
+    }
+
+    send(msg: string) {
+        try{
+            this.socket.send(msg);
+        } catch (err) {
+            Logger.log(LogLevel.ERROR, `could not send zmq message:\n${msg}`)
+        }
+    }
+
+    close() {
+        if (this.socket.closed === false) {
+            this.socket.close();
+        }
```

Comment:
depending on how socket.close() works internally you may not need the if statement. Generally close()-like operations are idempotent (they typically have this same if statement internally), you should double check.

Reply:
In our current version of zmq, if a socket is closed, and then `.close()` is used on it, zmq will throw an error and quit so I think it's safe to keep this

Reply:
wow, surprised they don't handle that, sad

Reply:
agreed :(